druntime/src/core/stdc/stdatomic.d

```diff
@@ -0,0 +1,1124 @@
+/**
+ * A D implementation of the C stdatomic.h header.
+ *
+ * $(NOTE If it compiles it should produce similar assembly to the system C toolchain
+ *   and should not introduce when optimizing unnecessary behaviors,
+ *   if you do not care about this guarantee use the _impl suffix.)
+ *
+ * $(NOTE The D shared type qualifier is the closest to the _Atomic type qualifier from C. It may be changed from shared in the future.)
+ *
+ * Copyright: Copyright Richard (Rikki) Andrew Cattermole 2023.
+ * License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Authors:   Richard (Rikki) Andrew cattermole
+ * Source:    $(DRUNTIMESRC core/stdc/stdatomic.d)
+ */
+module core.stdc.stdatomic;
+import core.atomic : MemoryOrder;
+import core.internal.atomic;
+import core.stdc.config;
+import core.stdc.stdint;
+
+@safe nothrow @nogc:
+
+///
+enum memory_order
+{
+    /// No ordering provided
+    memory_order_relaxed = MemoryOrder.raw,
+    /// As per cppreference.com circa 2015 no compiler supports consume memory order and in practice it devolves to acquire.
+    memory_order_consume = MemoryOrder.acq,
+    /// Prevent reordering before operation
+    memory_order_acquire = MemoryOrder.acq,
+    /// Prevent reordering after operation
+    memory_order_release = MemoryOrder.rel,
+    /// Prevent reordering before and after operation
+    memory_order_acq_rel = MemoryOrder.acq_rel,
+    /// Prevent reordering before for read operations and after for writes.
+    memory_order_seq_cst = MemoryOrder.seq
+}
+
+///
+enum
+{
+    ///
+    __STDC_VERSION_STDATOMIC_H__ = 202311,
+
+    ///
+    ATOMIC_BOOL_LOCK_FREE = IsAtomicLockFree!bool ? 2 : 0,
+    ///
+    ATOMIC_CHAR_LOCK_FREE = IsAtomicLockFree!char ? 2 : 0,
+    ///
+    ATOMIC_CHAR16_T_LOCK_FREE = IsAtomicLockFree!wchar ? 2 : 0,
+    ///
+    ATOMIC_CHAR32_T_LOCK_FREE = IsAtomicLockFree!dchar ? 2 : 0,
+    ///
+    ATOMIC_WCHAR_T_LOCK_FREE = ATOMIC_CHAR16_T_LOCK_FREE,
+    ///
+    ATOMIC_SHORT_LOCK_FREE = IsAtomicLockFree!short ? 2 : 0,
+    ///
+    ATOMIC_INT_LOCK_FREE = IsAtomicLockFree!int ? 2 : 0,
+    ///
+    ATOMIC_LONG_LOCK_FREE = IsAtomicLockFree!c_long ? 2 : 0,
+    ///
+    ATOMIC_LLONG_LOCK_FREE = IsAtomicLockFree!ulong ? 2 : 0,
+    ///
+    ATOMIC_POINTER_LOCK_FREE = IsAtomicLockFree!(void*) ? 2 : 0,
+    ///
+    ATOMIC_CHAR8_T_LOCK_FREE = ATOMIC_CHAR_LOCK_FREE,
+}
+
+version (DigitalMars)
+{
+    alias atomic_signal_fence = atomic_signal_fence_impl; ///
+
+    // these all use inline assembly, so will unlikely produce the codegen a user will expect
+    version(none) {
+        alias atomic_flag_clear = atomic_flag_clear_impl; ///
+        alias atomic_flag_clear_explicit = atomic_flag_clear_explicit_impl; ///
```

Comment:
It's ok to be slow, but this should be a choice that someone makes, otherwise why do we need this module when we have ``core.atomic`` already?

Instead, the goal here is to provide the guarantee that this will be as close as possible to the system C compiler in terms of codegen and if it isn't, that's a bug. To ensure ported code is as close to the original as possible in terms of runtime behavior.